### PR TITLE
feat(hybrid_zero_optim): add pipeline memory balance

### PR DESCRIPTION
**Pull request by mwiacx**
### Motivation

Due to the different number of activations on different stages of the pipeline, the memory usage on different devices varies.

### Modification

Balance the memory load across ranks by delegating the optimizer parameters of the stage with heavy activation load to the optimizer of the stage with light activation load.

### CODE REVIEW

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 1]
```diff
@@ -439,23 +543,54 @@ def _sync_grad(self):
         self._param_store.reset_reduced_data_for_compute_norm()
 
         # accumulate gradient
+        proxy_gradinets = []
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py:546]: 这里gradinets拼写错了

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 2]
```diff
@@ -439,23 +543,54 @@ def _sync_grad(self):
...
-                    if param.grad is not None:
-                        if len(avg_gradients[group_id]) == param_idx:
-                            avg_gradients[group_id].append(param.grad)
-                        else:
-                            avg_gradients[group_id][param_idx].add_(param.grad)
-                        param_idx += 1
+                    assert param.grad is not None, "gradient of proxy parameter is None"
+                    proxy_gradinets.append(param.grad)
+
+        # send offload gradients to reciever
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py:578]: reciever拼写错了

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 3]
```diff
@@ -439,23 +543,54 @@ def _sync_grad(self):
...
+        if self._enable_memory_balance and self._memory_balance_role == 0:
+            flat_proxy_grads = flatten(proxy_gradinets)
+
+            dist.send(flat_proxy_grads, self._memory_balance_peer)
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py:582]: 为啥这里不需要先send_obj_meta呢

Comment (mwiacx) [reply internlm/solver/optimizer/hybrid_zero_optim.py:582]: grads的形状不需要接收，用param的形状就行

File: internlm/solver/optimizer/hybrid_zero_optim.py (hunk 2 above)
Comment (mwiacx) [reply internlm/solver/optimizer/hybrid_zero_optim.py]: done

File: internlm/solver/optimizer/hybrid_zero_optim.py (hunk 1 above)
Comment (mwiacx) [reply internlm/solver/optimizer/hybrid_zero_optim.py]: done

**Pull request closed without merging**